### PR TITLE
Add HA-like solver mode support

### PR DESCRIPTION
**Pull request by randomir**

### CODE REVIEW

File: factoring/interfaces.py [hunk 1]
```diff
@@ -74,7 +92,7 @@ def factor(P, use_saved_embedding=True):
     sample_time = time.time()
 
     # get QPU sampler
-    sampler = DWaveSampler()
+    sampler = DWaveSampler(solver_features=dict(online=True, name='DW_2000Q.*'))
```
Comment (randomir) [inline factoring/interfaces.py:95]: Note: in prod, when list of solvers is limited for each user, `name` regex can be left out -- `online=True` is enough.

**Review by arcondello — APPROVED** [review]

File: factoring/interfaces.py (hunk 1 above)
Comment (JoelPasvolsky) [inline factoring/interfaces.py:95]: The current naming for online solvers in prod is DW_2000Q_ suffixed by 2 or 3

File: factoring/interfaces.py [hunk 2]
```diff
@@ -16,17 +16,34 @@
...
 
 
+def qpu_ha(f):
+    """High-availability QPU wrapper: retry (with next solver) if active solver
+    goes offline.
+    """
+
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        while True:
```
Comment (JoelPasvolsky) [inline factoring/interfaces.py:38]: Do you not want to limit the number of attempts here?

**Review by JoelPasvolsky — APPROVED** [review]

File: factoring/interfaces.py (hunk 2 above)
Comment (randomir) [reply factoring/interfaces.py:38]: No. :)

`SolverOfflineError` exception is raised in an *extremely* rare case -- only if online-to-offline switch happened in the middle of your sampling. After retry we'll fetch the new online solver and retry with that one.

File: factoring/interfaces.py (hunk 1 above)
Comment (randomir) [reply factoring/interfaces.py:95]: That's fine, `DW_2000Q.*` would match them both.

**Review by tcnleung — APPROVED** [review]

File: factoring/interfaces.py (hunk 1 above)
Comment (JoelPasvolsky) [reply factoring/interfaces.py:95]: Yes, but also for example DW_2000Q_VFYC_2, and I don't know that we have pre-embeddings for them

File: factoring/interfaces.py (hunk 2 above)
Comment (JoelPasvolsky) [reply factoring/interfaces.py:38]: I meant, don't you want to stop and return an error if there are no online solvers?

File: factoring/interfaces.py (hunk 1 above)
Comment (randomir) [reply factoring/interfaces.py:95]: That's a very valid point :)

I'm not sure what's the list Leap users will see?

If VFYC is included, we can use `DW_2000Q_[12]` for regex.

File: factoring/interfaces.py (hunk 2 above)
Comment (randomir) [reply factoring/interfaces.py:38]: Generic `SolverError` is passed thru.

File: factoring/interfaces.py (hunk 1 above)
Comment (JoelPasvolsky) [reply factoring/interfaces.py:95]: According to Vlad and Pete, they will see only DW_2000Q_2 and DW_2000Q_3, that is what they just Slacked me now

Comment (randomir) [reply factoring/interfaces.py:95]: :+1:

That's why I would just get rid of the complete name part in prod. @tcnleung

**Pull request merged**